src/main/cpp/src/get_json_object.hpp

```diff
@@ -0,0 +1,1210 @@
+/*
+ * Copyright (c) 2024, NVIDIA CORPORATION.
+ *
+ * Licensed under the Apache License, Version 2.0 (the "License");
+ * you may not use this file except in compliance with the License.
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+
+#include "json_parser.hpp"
+
+#include <cudf/strings/string_view.hpp>
+#include <cudf/strings/strings_column_view.hpp>
+
+#include <thrust/optional.h>
+#include <thrust/pair.h>
+#include <thrust/scan.h>
+#include <thrust/tuple.h>
+
+#include <memory>
+#include <string>
+#include <vector>
+
+namespace spark_rapids_jni {
+
+/**
+ * path instruction type
+ */
+enum class path_instruction_type { SUBSCRIPT, WILDCARD, KEY, INDEX, NAMED };
+
+namespace detail {
+
+/**
+ * write JSON style
+ */
+enum class write_style { raw_style, quoted_style, flatten_style };
+
+/**
+ * path instruction
+ */
+struct path_instruction {
+  CUDF_HOST_DEVICE inline path_instruction(path_instruction_type _type) : type(_type) {}
+
+  path_instruction_type type;
+
+  // used when type is named type
+  cudf::string_view name;
+
+  // used when type is index
+  int index{-1};
+};
+
+/**
+ * JSON generator is used to write out JSON string.
+ * It's not a full featured JSON generator, because get json object
+ * outputs an array or single item. JSON object is wroten as a whole item.
+ */
+template <int max_json_nesting_depth = curr_max_json_nesting_depth>
+class json_generator {
```

Comment:
Nit: In general, we should not put the implementation code of a class in the header. We can forward declare this class then implement it in a source file with explicit instantiation. So in future work we can clean up this that way.

Reply:
In addition, if we don't need to use multiple value of `max_json_nesting_depth`, we just declare it as `constexpr` and then declare a normal `json_generator` class without template parameter.

Reply:
Will do in a follow-on.